cache/models/samples.js

```diff
@@ -539,20 +547,33 @@ function upsertOneParsedSample(sampleQueryBodyObj, parsedSample, isBulk, user) {
    * block and return an error. Otherwise, we return the sample.
    */
   return checkWritePermission(aspectName, userName, isBulk)
-  .then(() => Promise.all([
-    redisClient.hgetallAsync(subjKey),
-    redisClient.hgetallAsync(sampleKey),
-  ])
-  .then((responses) => {
-    [subject, sample] = responses;
-
-    if (!subject || subject.isPublished === 'false') {
-      handleUpsertError(constants.objectType.subject, isBulk, sampleName);
+  .then(() => redisClient.hgetallAsync(sampleKey))
+  .then((response) => {
+    sample = response;
+
+    if (!sample) { // new sample. check subject/asp is published from db
+      const subjWhereObj = {
+        where: { absolutePath: { [Op.iLike]: absolutePath } },
+      };
+      const aspWhereObj = { where: { name: { [Op.iLike]: aspectName } } };
+      return Promise.join(
+        db.Subject.findOne(subjWhereObj),
+        db.Aspect.findOne(aspWhereObj)
+      )
```

Comment:
> Seems like going to the db on every upsert defeats the purpose of storing the extra info in redis.
We are not doing this on every upsert. We do this only when the sample is new. I remember this is what we decided for new samples flow. And we no longer need to get aspects in bulkUpsertByName for each sample. I think we should sync up offline on this.

Reply:
> We are not doing this on every upsert. We do this only when the sample is new. 
Ah, ok, I missed that - I remember that discussion now. Disregard this comment, I only thought about doing the operations outside of the sample loop because I thought they would be happening every time.

Though the idea of attaching the tags and doing the status calculation outside the loop might still be worth it, as it would allow us to send all the redis operations for the upsert in a single batch. Thoughts?

@iamigo

Reply:
I think this is good idea - Moving the tags and status lookups outside the sample loop. And if we are thinking on those lines, maybe we should move checkPermissions also out of samples loop too?
If we plan to go that route, I think that should be another story because it will involve moving critical code around. I don't want that optimization to get lost in the realtime format changes.